File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.cpp
class Solution {
public:
    int ans=0;
    vector<vector<int>> adj;
    bitset<30001> visited=0;
    inline void build_adj(int n, vector<vector<int>>& edges, vector<int>& values)
    {
        for(auto& e: edges){
            int i=e[0], j=e[1];
            adj[i].push_back(j);
            adj[j].push_back(i);
        }
    }
    inline long long dfs(int i, vector<int>& values, int k){
        visited[i]=1;
        long long sum=values[i];
        for (int j: adj[i]){
            if (visited[j]) continue;
            sum+=dfs(j, values, k);
            sum%=k;
        }
        if (sum%k==0){
            ans++;
            return 0;
        }
        return sum;
    }
    
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) 
    {
        adj.resize(n);
    
        build_adj(n, edges, values);

        dfs(0, values, k);

        return ans;
    }
};



auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.py
from collections import defaultdict
from typing import List

class Solution:
    def __init__(self):
        self.ans = 0
        self.adj = defaultdict(list)
        self.visited = [False] * 30001

    def build_adj(self, edges: List[List[int]]) -> None:
        for i, j in edges:
            self.adj[i].append(j)
            self.adj[j].append(i)

    def dfs(self, i: int, values: List[int], k: int) -> int:
        self.visited[i] = True
        total = values[i]
        for j in self.adj[i]:
            if not self.visited[j]:
                total += self.dfs(j, values, k)
                total %= k
        if total % k == 0:
            self.ans += 1
            return 0
        return total

    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        self.build_adj(edges)
        self.dfs(0, values, k)
        return self.ans
